dev/usb/class/cdcserial/cdcserial.c

```diff
@@ -0,0 +1,300 @@
+        // real modem, we're probably okay to drop them on the floor and reply with
+        // an acknowledgement.
+        if (setup->length) {  // Has data phase?
+            if (setup->request_type & 0x80) {
+                // We have to send data?
+                usbc_ep0_send(buf, setup->length, 64);
+            } else {
+                usbc_ep0_recv(buf, setup->length, NULL);
+                usbc_ep0_ack();
+            }
+        } else {
+            switch (setup->request) {
+                case CDC_REQ_SEND_CMD:
+                case CDC_REQ_GET_RESP:
+                case CDC_REQ_SET_COMM_FEATURE:
+                case CDC_REQ_GET_COMM_FEATURE:
+                case CDC_REQ_CLEAR_COMM_FEATURE:
+                case CDC_REQ_SET_LINE_CODING:
+                case CDC_REQ_GET_LINE_CODING:
+                case CDC_REQ_SET_CONTROL_LINE_STATE:
+                case CDC_REQ_SEND_BREAK:
+                    // Ack any command that we understand.
+                    usbc_ep0_ack();
+                    break;
+            }
+        }
+    }
+
+    return NO_ERROR;
+}
+
+status_t cdcserial_init(void)
+{
+    usb_register_callback(&usb_register_cb, NULL);
+    return NO_ERROR;
+}
+
+void cdcserial_create_channel(int data_ep_addr, int ctrl_ep_addr)
+{
+    ctrl_if_descriptor[CTRL_IN_EP_ADDR_OFFSET]  = ctrl_ep_addr | 0x80;
+    data_if_descriptor[DATA_IN_EP_ADDR_OFFSET]  = data_ep_addr | 0x80;
+    data_if_descriptor[DATA_OUT_EP_ADDR_OFFSET] = data_ep_addr;
+
+    ctrl_if_descriptor[LEADER_EP_NUMBER_OFFSET] =
+        usb_get_current_iface_num_lowspeed();
+    ctrl_if_descriptor[FOLLOWER_EP_NUMBER_OFFSET] =
+        usb_get_current_iface_num_lowspeed() + 1;
+
+    usb_append_interface_lowspeed(ctrl_if_descriptor, sizeof(ctrl_if_descriptor));
+    usb_append_interface_lowspeed(data_if_descriptor, sizeof(data_if_descriptor));
+
+    ctrl_if_descriptor[LEADER_EP_NUMBER_OFFSET] =
+        usb_get_current_iface_num_highspeed();
+    ctrl_if_descriptor[FOLLOWER_EP_NUMBER_OFFSET] =
+        usb_get_current_iface_num_highspeed() + 1;
+
+    usb_append_interface_highspeed(ctrl_if_descriptor, sizeof(ctrl_if_descriptor));
+    usb_append_interface_highspeed(data_if_descriptor, sizeof(data_if_descriptor));
+
+    registered_bulk_eps_in |= (0x1 << data_ep_addr);
+    registered_bulk_eps_out |= (0x1 << data_ep_addr);
+    registered_intr_eps_in |= (0x1 << ctrl_ep_addr);
+}
+
+static status_t usb_xmit_cplt_cb(ep_t endpoint, usbc_transfer_t *t)
+{
+    event_signal(&txevt, false);
+    return 0;
+}
+
+static status_t usb_recv_cplt_cb(ep_t endpoint, usbc_transfer_t *t)
+{
+    event_signal(&rxevt, false);
+    return 0;
+}
+
+// Write len bytes to the CDC Serial Virtual Com Port.
+status_t cdcserial_write(size_t len, uint8_t *buf)
+{
+    LTRACEF("len = %d, buf = %p\n", len, buf);
+
+    DEBUG_ASSERT(buf);
+
+    if (!usb_online) {
+        return ERR_NOT_READY;
+    }
+
+    usbc_transfer_t transfer = {
```

Comment:
as i pointed out in other change, i think you want this to be static because the usbc code will hold onto a pointer to this transfer and pass it back in the callback.